main/src/ca/uwaterloo/flix/api/lsp/provider/SemanticTokensProvider.scala

```diff
@@ -0,0 +1,99 @@
+package ca.uwaterloo.flix.api.lsp.provider
+
+import ca.uwaterloo.flix.api.lsp.{Entity, Index}
+import ca.uwaterloo.flix.language.ast.SourceLocation
+import ca.uwaterloo.flix.language.ast.TypedAst.{Expression, Root}
+import org.json4s.JsonAST.JObject
+import org.json4s.JsonDSL._
+
+import scala.collection.mutable.ArrayBuffer
+
+object SemanticTokensProvider {
+  /**
+   * The token types currently supported by the provider.
+   *
+   * Note: The intermediate TypeScript server is responsible for communicating the legend
+   * to the client. The TS server must list the token types in the same order as is used
+   * here in the Scala code. The reason for this is that the response is encoded as a
+   * sequence of integers, and the integer representing a token type is the type's
+   * index in the legend.
+   */
+  private object TokenType extends Enumeration {
+    type TokenType = Value
+
+    val Number, Str = Value
+  }
+
+  /**
+   * The token modifiers currently supported by the provider.
+   *
+   * Note: See the remark in the documentation for SemanticTokenType; the same
+   * constraint applies to modifiers: The TS server must list the modifiers
+   * in the same order as is used here in the Scala code.
+   */
+  private object TokenModifier extends Enumeration {
+    type TokenModifier = Value
+
+    val Declaration, Definition = Value
+  }
+
+  private case class SemanticToken(loc: SourceLocation,
+                                   tokenType: SemanticTokensProvider.TokenType.TokenType,
+                                   tokenModifiers: List[SemanticTokensProvider.TokenModifier.TokenModifier])
+
+  def provideSemanticTokens(uri: String)(implicit index: Index, root: Root): JObject = {
+    val entities = index.query(uri)
+    val semanticTokens = entities.flatMap(getSemanticTokens)
+    val encoding = encodeSemanticTokens(semanticTokens)
+    val result = ("data" -> encoding)
+    ("status" -> "success") ~ ("result" -> result)
+  }
+
+  private def getSemanticTokens(entity: Entity): List[SemanticToken] = entity match {
+    case Entity.Exp(e) => e match {
```

Comment:
This approach looks right to me. I think you should be able to color everything but keywords (because they are not in the AST).

Do you know if SemanticTokens are additive? Are they layered on top of the grammer or do they replace it?

Reply:
It seems they are additive:

https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide

```
Semantic tokenization allows language servers to provide additional token information based on the language server's knowledge on how to resolve symbols in the context of a project. Themes can opt in to use semantic tokens to improve and refine the syntax highlighting from grammars.
```

I think this means that over time we want to have a simple grammar that provides basic syntax coloring (e.g. it could just be keywords) and then we can use semantic tokens to fill in the rest! :)

Reply:
From my experiments it seems that semantic highlighting takes precedence over highlighting provided by the `flix.tmLanguage.json` grammar file. For example, if the grammar file classifies a token as a variable and the language server classifies the token as a function, then the token will be highlighted as a function.

Since the language server knows the semantics of the file, it can provide more accurate highlighting than the simple grammar file. On the other hand, highlighting via the grammar file is faster than doing a round trip to the LS. So there's a tradeoff between speed and accuracy.

Also, from my experiments the LS classifies e.g. `Array.toList` as a function name (`Def` in compiler lingo?), but (to my mind at least) only `toList` should be classified as a function name, and `Array` should be classified as a namespace (like we do currently with the grammar file).

I think it's a nice idea to let semantic tokens drive the syntax highlighting. But probably some trial and error is necessary to get it right (and fast). 🙂

Reply:
OK, this supports my intuition that we should have a base grammar file and then leave more complicated stuff to semantic tokens.